File: Classes/MediaListModel.m
// Copyright 2016 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "MediaListModel.h"

#import <GoogleCast/GoogleCast.h>

#import "MediaItem.h"

NSString *const kMediaKeyPosterURL = @"posterUrl";
NSString *const kMediaKeyDescription = @"description";

static NSString *const kKeyCategories = @"categories";
static NSString *const kKeyMP4BaseURL = @"mp4";
static NSString *const kKeyImagesBaseURL = @"images";
static NSString *const kKeyTracksBaseURL = @"tracks";
static NSString *const kKeySources = @"sources";
static NSString *const kKeyVideos = @"videos";
static NSString *const kKeyArtist = @"artist";
static NSString *const kKeyBaseURL = @"baseUrl";
static NSString *const kKeyContentID = @"contentId";
static NSString *const kKeyDescription = @"description";
static NSString *const kKeyID = @"id";
static NSString *const kKeyImageURL = @"image-480x270";
static NSString *const kKeyItems = @"items";
static NSString *const kKeyLanguage = @"language";
static NSString *const kKeyMimeType = @"mime";
static NSString *const kKeyName = @"name";
static NSString *const kKeyPosterURL = @"image-780x1200";
static NSString *const kKeyStreamType = @"streamType";
static NSString *const kKeyStudio = @"studio";
static NSString *const kKeySubtitle = @"subtitle";
static NSString *const kKeySubtype = @"subtype";
static NSString *const kKeyTitle = @"title";
static NSString *const kKeyTracks = @"tracks";
static NSString *const kKeyType = @"type";
static NSString *const kKeyURL = @"url";
static NSString *const kKeyDuration = @"duration";

static NSString *const kDefaultVideoMimeType = @"video/mp4";
static NSString *const kDefaultTrackMimeType = @"text/vtt";

static NSString *const kTypeAudio = @"audio";
static NSString *const kTypePhoto = @"photos";
static NSString *const kTypeVideo = @"videos";

static NSString *const kTypeLive = @"live";

static const NSInteger kThumbnailWidth = 480;
static const NSInteger kThumbnailHeight = 720;

static const NSInteger kPosterWidth = 780;
static const NSInteger kPosterHeight = 1200;

@interface MediaListModel ()<NSURLConnectionDelegate,
                             NSURLConnectionDataDelegate> {
  NSURLRequest *_request;
  NSURLConnection *_connection;
  NSMutableData *_responseData;
  NSInteger _responseStatus;
  GCKMediaTextTrackStyle *_trackStyle;
  MediaItem *_rootItem;
}

@property(nonatomic, readwrite) NSString *title;
@property(nonatomic, assign, readwrite) BOOL loaded;

@end

@implementation MediaListModel {
  /** Storage for the list of Media objects. */
  NSArray *_medias;
}

- (instancetype)init {
  if (self = [super init]) {
    _trackStyle = [GCKMediaTextTrackStyle createDefault];
  }
  return self;
}

- (void)loadFromURL:(NSURL *)url {
  _rootItem = nil;
  _request = [NSURLRequest requestWithURL:url];
  _connection = [NSURLConnection connectionWithRequest:_request delegate:self];
  _responseData = nil;
  [_connection start];
  GCKLog(@"loading media list from URL %@", url);
}

- (void)cancelLoad {
  if (_request) {
    [_connection cancel];
    _request = nil;
    _connection = nil;
    _responseData = nil;
  }
}

- (MediaItem *)rootItem {
  return _rootItem;
}

#pragma mark - NSURLConnectionDelegate

- (void)connection:(NSURLConnection *)connection
  didFailWithError:(NSError *)error {
  _request = nil;
  _responseData = nil;
  _connection = nil;
  GCKLog(@"httpRequest failed with %@", error);
  [self.delegate mediaListModel:self didFailToLoadWithError:error];
}

#pragma mark - NSURLConnectionDataDelegate

- (void)connection:(NSURLConnection *)connection
didReceiveResponse:(NSURLResponse *)response {
  if ([response respondsToSelector:@selector(statusCode)]) {
    _responseStatus = ((NSHTTPURLResponse *)response).statusCode;
  }
}

- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
  if (!_responseData) {
    _responseData = [[NSMutableData alloc] init];
  }
  [_responseData appendData:data];
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection {
  GCKLog(@"httpRequest completed with %ld", (long)_responseStatus);
  if (_responseStatus == 200) {
    NSError *error;
    NSDictionary *jsonData =
        [NSJSONSerialization JSONObjectWithData:_responseData
                                        options:kNilOptions
                                          error:&error];
    _rootItem = [self decodeMediaTreeFromJSON:jsonData];
    self.loaded = YES;
    [self.delegate mediaListModelDidLoad:self];
  } else {
    NSError *error = [[NSError alloc] initWithDomain:@"HTTP"
                                                code:_responseStatus
                                            userInfo:nil];
    [self.delegate mediaListModel:self didFailToLoadWithError:error];
  }
}

#pragma mark - JSON decoding

- (MediaItem *)decodeMediaTreeFromJSON:(NSDictionary *)json {
  MediaItem *rootItem =
      [[MediaItem alloc] initWithTitle:nil imageURL:nil parent:nil];

  NSArray *categories = [json gck_arrayForKey:kKeyCategories];
  for (NSDictionary *categoryElement in categories) {
    if (![categoryElement isKindOfClass:[NSDictionary class]]) continue;

    NSDictionary *category = (NSDictionary *)categoryElement;

    NSArray *mediaList = [category gck_arrayForKey:kKeyVideos];
    if (mediaList && [mediaList isKindOfClass:[NSArray class]]) {
      self.title = [category gck_stringForKey:kKeyName];
      // Pick the MP4 files only
      NSString *videosBaseURLString =
          [category gck_stringForKey:kKeyMP4BaseURL];
      NSURL *videosBaseURL = [NSURL URLWithString:videosBaseURLString];
      NSString *imagesBaseURLString =
          [category gck_stringForKey:kKeyImagesBaseURL];
      NSURL *imagesBaseURL = [NSURL URLWithString:imagesBaseURLString];
      NSString *tracksBaseURLString =
          [category gck_stringForKey:kKeyTracksBaseURL];
      NSURL *tracksBaseURL = [NSURL URLWithString:tracksBaseURLString];
      [self decodeItemListFromArray:mediaList
                           intoItem:rootItem
                        videoFormat:kKeyMP4BaseURL
                      videosBaseURL:videosBaseURL
                      imagesBaseURL:imagesBaseURL
                      tracksBaseURL:tracksBaseURL];
      break;
    }
  }

  return rootItem;
}

- (NSURL *)buildURLWithString:(NSString *)string baseURL:(NSURL *)baseURL {
  if (!string) return nil;

  if ([string hasPrefix:@"http://"] || [string hasPrefix:@"https://"]) {
    return [NSURL URLWithString:string];
  } else {
    return [NSURL URLWithString:string relativeToURL:baseURL];
  }
}

- (void)decodeItemListFromArray:(NSArray *)array
                       intoItem:(MediaItem *)item
                    videoFormat:(NSString *)videoFormat
                  videosBaseURL:(NSURL *)videosBaseURL
                  imagesBaseURL:(NSURL *)imagesBaseURL
                  tracksBaseURL:(NSURL *)tracksBaseURL {
  for (NSDictionary *element in array) {
    if (![element isKindOfClass:[NSDictionary class]]) continue;

    NSDictionary *dict = (NSDictionary *)element;

    NSString *title = [dict gck_stringForKey:kKeyTitle];

    GCKMediaMetadata *metadata = [[GCKMediaMetadata alloc]
        initWithMetadataType:GCKMediaMetadataTypeMovie];
    [metadata setString:title forKey:kGCKMetadataKeyTitle];

    NSString *mimeType = nil;
    NSURL *url = nil;
    NSArray *sources = [dict gck_arrayForKey:kKeySources];
    for (id sourceElement in sources) {
      if (![sourceElement isKindOfClass:[NSDictionary class]]) continue;

      NSDictionary *sourceDict = (NSDictionary *)sourceElement;

      NSString *type = [sourceDict gck_stringForKey:kKeyType];
      if ([type isEqualToString:videoFormat]) {
        mimeType = [sourceDict gck_stringForKey:kKeyMimeType];
        NSString *urlText = [sourceDict gck_stringForKey:kKeyURL];
        url = [self buildURLWithString:urlText baseURL:videosBaseURL];
        break;
      }
    }

    NSString *imageURLString = [dict gck_stringForKey:kKeyImageURL];
    NSURL *imageURL =
        [self buildURLWithString:imageURLString baseURL:imagesBaseURL];
    if (imageURL) {
      [metadata addImage:[[GCKImage alloc] initWithURL:imageURL
                                                 width:kThumbnailWidth
                                                height:kThumbnailHeight]];
    }

    NSString *posterURLText = [dict gck_stringForKey:kKeyPosterURL];
    NSURL *posterURL =
        [self buildURLWithString:posterURLText baseURL:imagesBaseURL];
    if (posterURL) {
      [metadata setString:posterURL.absoluteString forKey:kMediaKeyPosterURL];
      [metadata addImage:[[GCKImage alloc] initWithURL:posterURL
                                                 width:kPosterWidth
                                                height:kPosterHeight]];
    }

    NSString *description = [dict gck_stringForKey:kKeySubtitle];
    if (description) {
      [metadata setString:description forKey:kMediaKeyDescription];
    }

    NSMutableArray *mediaTracks = nil;

    NSString *studio = [dict gck_stringForKey:kKeyStudio];
    [metadata setString:studio forKey:kGCKMetadataKeyStudio];

    NSInteger duration = [dict gck_integerForKey:kKeyDuration];

    mediaTracks = [[NSMutableArray alloc] init];

    NSArray *tracks = [dict gck_arrayForKey:kKeyTracks];
    for (id trackElement in tracks) {
      if (![trackElement isKindOfClass:[NSDictionary class]]) continue;

      NSDictionary *trackDict = (NSDictionary *)trackElement;

      NSInteger identifier = [trackDict gck_integerForKey:kKeyID];
      NSString *name = [trackDict gck_stringForKey:kKeyName];
      NSString *typeString = [trackDict gck_stringForKey:kKeyType];
      NSString *subtypeString = [trackDict gck_stringForKey:kKeySubtype];
      NSString *contentID = [trackDict gck_stringForKey:kKeyContentID];
      NSString *language = [trackDict gck_stringForKey:kKeyLanguage];

      NSURL *url = [self buildURLWithString:contentID baseURL:tracksBaseURL];

      GCKMediaTrack *mediaTrack = [[GCKMediaTrack alloc]
          initWithIdentifier:identifier
           contentIdentifier:url.absoluteString
                 contentType:kDefaultTrackMimeType
                        type:[self trackTypeFrom:typeString]
                 textSubtype:[self textTrackSubtypeFrom:subtypeString]
                        name:name
                languageCode:language
                  customData:nil];
      [mediaTracks addObject:mediaTrack];
    }
    if (mediaTracks.count == 0) {
      mediaTracks = nil;
    }

    GCKMediaInformation *mediaInfo = [[GCKMediaInformation alloc]
        initWithContentID:url.absoluteString
               streamType:GCKMediaStreamTypeBuffered
              contentType:mimeType
                 metadata:metadata
           streamDuration:duration
              mediaTracks:mediaTracks
           textTrackStyle:_trackStyle
               customData:nil];

    MediaItem *childItem =
        [[MediaItem alloc] initWithMediaInformation:mediaInfo parent:item];
    [item.items addObject:childItem];
  }
}

- (GCKMediaTrackType)trackTypeFrom:(NSString *)string {
  if ([string isEqualToString:@"audio"]) {
    return GCKMediaTrackTypeAudio;
  }
  if ([string isEqualToString:@"text"]) {
    return GCKMediaTrackTypeText;
  }
  if ([string isEqualToString:@"video"]) {
    return GCKMediaTrackTypeVideo;
  }
  return GCKMediaTrackTypeUnknown;
}

- (GCKMediaTextTrackSubtype)textTrackSubtypeFrom:(NSString *)string {
  if ([string isEqualToString:@"captions"]) {
    return GCKMediaTextTrackSubtypeCaptions;
  }
  if ([string isEqualToString:@"chapters"]) {
    return GCKMediaTextTrackSubtypeChapters;
  }
  if ([string isEqualToString:@"descriptions"]) {
    return GCKMediaTextTrackSubtypeDescriptions;
  }
  if ([string isEqualToString:@"metadata"]) {
    return GCKMediaTextTrackSubtypeMetadata;
  }
  if ([string isEqualToString:@"subtitles"]) {
    return GCKMediaTextTrackSubtypeSubtitles;
  }

  return GCKMediaTextTrackSubtypeUnknown;
}

@end

File: CastVideos-swift/MediaListModel.swift
// Copyright 2016 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import Foundation
import GoogleCast

/** A key for the URL of the media item's poster (large image). */
let kMediaKeyPosterURL: String = "posterUrl"
/** A key for the media item's extended description. */
let kMediaKeyDescription: String = "description"

let kKeyCategories: String = "categories"

let kKeyMP4BaseURL: String = "mp4"

let kKeyImagesBaseURL: String = "images"

let kKeyTracksBaseURL: String = "tracks"

let kKeySources: String = "sources"

let kKeyVideos: String = "videos"

let kKeyArtist: String = "artist"

let kKeyBaseURL: String = "baseUrl"

let kKeyContentID: String = "contentId"

let kKeyDescription: String = "description"

let kKeyID: String = "id"

let kKeyImageURL: String = "image-480x270"

let kKeyItems: String = "items"

let kKeyLanguage: String = "language"

let kKeyMimeType: String = "mime"

let kKeyName: String = "name"

let kKeyPosterURL: String = "image-780x1200"

let kKeyStreamType: String = "streamType"

let kKeyStudio: String = "studio"

let kKeySubtitle: String = "subtitle"

let kKeySubtype: String = "subtype"

let kKeyTitle: String = "title"

let kKeyTracks: String = "tracks"

let kKeyType: String = "type"

let kKeyURL: String = "url"

let kKeyDuration: String = "duration"

let kDefaultVideoMimeType: String = "video/mp4"

let kDefaultTrackMimeType: String = "text/vtt"

let kTypeAudio: String = "audio"

let kTypePhoto: String = "photos"

let kTypeVideo: String = "videos"

let kTypeLive: String = "live"

let kThumbnailWidth: Int = 480

let kThumbnailHeight: Int = 720

let kPosterWidth: Int = 780

let kPosterHeight: Int = 1200

/**
 * The delegate protocol for receiving notifications from the model.
 */
protocol MediaListModelDelegate: NSObjectProtocol {
  /**
   * Called when the media list has loaded.
   *
   * @param list The media list.
   */
  func mediaListModelDidLoad(_ list: MediaListModel)
  /**
   * Called when the media list has failed to load.
   *
   * @param list The media list.
   * @param error The error.
   */

  func mediaListModel(_ list: MediaListModel, didFailToLoadWithError error: Error?)
}

/**
 * An object representing a hierarchy of media items.
 */
class MediaListModel: NSObject, NSURLConnectionDelegate, NSURLConnectionDataDelegate {
  var request: URLRequest!
  var connection: NSURLConnection!
  var responseData: Data!
  var responseStatus: Int = 0
  var trackStyle: GCKMediaTextTrackStyle!
  /* The root item (top-level group). */
  private(set) public var rootItem: MediaItem!

  /** A delegate for receiving notifications from the model. */
  weak var delegate: MediaListModelDelegate?
  /** A flag indicating whether the model has been loaded. */
  private(set) public var isLoaded: Bool = false
  /** The title of the media list. */
  private(set) var title: String = ""

  /** Storage for the list of Media objects. */
  var medias = [Any]()

  /**
   * Begins loading the model from the given URL. The delegate will be messaged
   * when the load
   * completes or fails.
   *
   * @param url The URL of the JSON file describing the media hierarchy.
   */

  func load(from url: URL) {
    self.rootItem = nil
    self.request = URLRequest(url: url)
    self.connection = NSURLConnection(request: self.request, delegate: self)
    self.responseData = nil
    self.connection.start()
    GCKLog("loading media list from URL %@", url)
  }



  override init() {
    super.init()

    self.trackStyle = GCKMediaTextTrackStyle.createDefault()

  }

  func cancelLoad() {
    if self.request != nil {
      self.connection.cancel()
      self.request = nil
      self.connection = nil
      self.responseData = nil
    }
  }
  // MARK: - NSURLConnectionDelegate

  func connection(_ connection: NSURLConnection, didFailWithError error: Error) {
    self.request = nil
    self.responseData = nil
    self.connection = nil
    GCKLog("httpRequest failed with %@", error)
    self.delegate?.mediaListModel(self, didFailToLoadWithError: error)
  }
  // MARK: - NSURLConnectionDataDelegate

  func connection(_ connection: NSURLConnection, didReceive response: URLResponse) {
    if response.responds(to: #selector(self.statusCode)) {
      self.responseStatus = (response as! HTTPURLResponse).statusCode
    }
  }

  func connection(_ connection: NSURLConnection, didReceive data: Data) {
    if !(self.responseData != nil) {
      self.responseData = Data()
    }
    self.responseData.append(data)
  }

  func connectionDidFinishLoading(_ connection: NSURLConnection) {
    GCKLog("httpRequest completed with %ld", Int(self.responseStatus))
    if self.responseStatus == 200 {
      var jsonData: [AnyHashable: Any]? = (try? JSONSerialization.jsonObject(withData: self.responseData, options: JSONSerialization.ReadingOptions.MutableContainers))
      self.rootItem = self.decodeMediaTree(fromJSON: jsonData!)
      self.isLoaded = true
      self.delegate?.mediaListModelDidLoad(self)
    }
    else {
      let error = NSError.init(domain: "HTTP", code: self.responseStatus, userInfo: nil)
      self.delegate?.mediaListModel(self, didFailToLoadWithError: error)
    }
  }
  // MARK: - JSON decoding

  func decodeMediaTree(fromJSON json: [AnyHashable: Any]) -> MediaItem {
    var rootItem = MediaItem(title: nil, imageURL: nil, parent: nil)
    var categories: [Any] = json.gck_array(forKey: kKeyCategories)
    for categoryElement: [AnyHashable: Any] in categories {
      if !(categoryElement is [AnyHashable: Any]) {
        continue
      }
      var category: [AnyHashable: Any]? = (categoryElement as? [AnyHashable: Any])
      var mediaList: [Any]? = category?.gck_array(forKey: kKeyVideos)
      if mediaList && (mediaList? is [Any]) {
        self.title = category?.gck_string(forKey: kKeyName)
        // Pick the MP4 files only
        var videosBaseURLString: String? = category?.gck_string(forKey: kKeyMP4BaseURL)
        var videosBaseURL = URL(string: videosBaseURLString)
        var imagesBaseURLString: String? = category?.gck_string(forKey: kKeyImagesBaseURL)
        var imagesBaseURL = URL(string: imagesBaseURLString)
        var tracksBaseURLString: String? = category?.gck_string(forKey: kKeyTracksBaseURL)
        var tracksBaseURL = URL(string: tracksBaseURLString)
        self.decodeItemList(fromArray: mediaList, into: rootItem, videoFormat: kKeyMP4BaseURL, videosBaseURL: videosBaseURL, imagesBaseURL: imagesBaseURL, tracksBaseURL: tracksBaseURL)
        break
      }
    }
    return rootItem
  }

  func buildURL(with string: String?, baseURL: URL) -> URL? {
    guard let string = string else { return nil}
    if string.hasPrefix("http://") || string.hasPrefix("https://") {
      return URL(string: string)!
    }
    else {
      return URL(string: string, relativeTo: baseURL)!
    }
  }

  func decodeItemList(fromArray array: [Any], into item: MediaItem, videoFormat: String, videosBaseURL: URL, imagesBaseURL: URL, tracksBaseURL: URL) {
    for element in array {
      if let dict = element as? NSDictionary {
        var metadata = GCKMediaMetadata(metadataType: .movie)
        if let title = dict.gck_string(forKey: kKeyTitle) {
          metadata.setString(title, forKey: kGCKMetadataKeyTitle)
        }
        var mimeType: String? = nil
        var url: URL? = nil
        var sources: [Any] = dict.gck_array(forKey: kKeySources) ?? []
        for sourceElement in sources {
          if let sourceDict = sourceElement as? NSDictionary {
            var type: String? = sourceDict.gck_string(forKey: kKeyType)
            if (type == videoFormat) {
              mimeType = sourceDict.gck_string(forKey: kKeyMimeType)
              var urlText: String? = sourceDict.gck_string(forKey: kKeyURL)
              url = self.buildURL(with: urlText, baseURL: videosBaseURL)
              break
            }
          }
          var imageURLString: String? = dict.gck_string(forKey: kKeyImageURL)
          var imageURL = self.buildURL(with: imageURLString, baseURL: imagesBaseURL)
          if let imageURL = imageURL {
            metadata.addImage(GCKImage(url: imageURL, width: kThumbnailWidth, height: kThumbnailHeight))
          }
          var posterURLText: String? = dict.gck_string(forKey: kKeyPosterURL)
          var posterURL = self.buildURL(with: posterURLText, baseURL: imagesBaseURL)
          if let posterURL = posterURL {
            metadata.setString(posterURL.absoluteString, forKey: kMediaKeyPosterURL)
            metadata.addImage(GCKImage(url: posterURL, width: kPosterWidth, height: kPosterHeight))
          }
          if let description = dict.gck_string(forKey: kKeySubtitle) {
            metadata.setString(description, forKey: kMediaKeyDescription)
          }
          var mediaTracks: [Any]? = nil
          if let studio = dict.gck_string(forKey: kKeyStudio) {
            metadata.setString(studio, forKey: kGCKMetadataKeyStudio)
          }
          var duration: Int? = dict.gck_integer(forKey: kKeyDuration)
          mediaTracks = [GCKMediaTrack]()
          if let tracks = dict.gck_array(forKey: kKeyTracks) {
            for trackElement: Any in tracks {
              guard let trackDict = trackElement as? NSDictionary else { continue }
              var identifier = trackDict.gck_integer(forKey: kKeyID)
              var name: String? = trackDict.gck_string(forKey: kKeyName)
              var typeString: String? = trackDict.gck_string(forKey: kKeyType)
              var subtypeString: String? = trackDict.gck_string(forKey: kKeySubtype)
              var contentID: String? = trackDict.gck_string(forKey: kKeyContentID)
              var language: String? = trackDict.gck_string(forKey: kKeyLanguage)
              var url = self.buildURL(with: contentID, baseURL: tracksBaseURL)
              var mediaTrack = GCKMediaTrack(identifier: identifier, contentIdentifier: url?.absoluteString, contentType: kDefaultTrackMimeType, type: self.trackType(from: typeString!), textSubtype: self.textTrackSubtype(from: subtypeString!), name: name, languageCode: language, customData: nil)
              mediaTracks?.append(mediaTrack)
            }
          }

          if mediaTracks?.count == 0 {
            mediaTracks = nil
          }
          var mediaInfo = GCKMediaInformation(contentID: url!.absoluteString, streamType: .buffered, contentType: mimeType!, metadata: metadata, streamDuration: TimeInterval(duration!), mediaTracks: mediaTracks as! [GCKMediaTrack]?, textTrackStyle: self.trackStyle, customData: nil)
          var childItem = MediaItem(mediaInformation: mediaInfo, parent: item)
          item.items.append(childItem)

        }
      }
    }
  }

  func trackType(from string: String) -> GCKMediaTrackType {
    if (string == "audio") {
      return .audio
    }
    if (string == "text") {
      return .text
    }
    if (string == "video") {
      return .video
    }
    return .unknown
  }

  func textTrackSubtype(from string: String) -> GCKMediaTextTrackSubtype {
    if (string == "captions") {
      return .captions
    }
    if (string == "chapters") {
      return .chapters
    }
    if (string == "descriptions") {
      return .descriptions
    }
    if (string == "metadata") {
      return .metadata
    }
    if (string == "subtitles") {
      return .subtitles
    }
    return .unknown
  }
}
